[test] Apply maximal flags to tx_valid tests and minimal flags to tx_invalid tests

jl2012 opened the pull request:
The first commit inverts the meaning of verifyFlags for tx_valid tests, as flags being excluded. All flags are applied by default, except those found in verifyFlags. This makes sure that a new or existing flag won't invalidate a tx by accident.

The second commit reduces the number of validation flags used for tx_invalid tests, to a minimally required set to fail a test. This makes sure that a tx failed due to the tested flags, not unexpected effects of some other flags. It also uses "BADTX" to indicate tests not passing CheckTransaction(), vs. those failing script execution.

(If a test is expected to fail due to multiple independent flags, multiple tests should be used)

The third commit verifies that the flags excluded in tx_valid and included in tx_invalid are indeed the minimal set. In tx_valid, it adds back the excluded flags individually and expects it to fail. In tx_invalid, it removes the included flags individually and expects it to pass.

This process helped me to identify and fix some buggy tests:
- Remove unnecessary OP_1 at the end of most OP_CLTV and OP_CSV tx_valid tests, so there is no need to exclude CLEANSTACK
- An OP_CSV tx_valid test missed an OP_ADD, and is added back
- 2 witness tests were found with empty vout, so they failed due to CheckTransaction(), not script tests. Corrected by filling in proper vout.

jl2012 committed:
Apply maximal validation flags to tx_valid tests

- Apply all validation flags by default
- Invert the meaning of verifyFlags as flags being excluded
- Remove unnecessary OP_1 at the end of most OP_CLTV and OP_CSV tests
- Fix an OP_CSV test with an OP_ADD missing

jl2012 committed:
Apply minimal validation flags to tx_invalid tests

- Reduce the number of validation flags used, to a minimally required set to fail a test
- Use "BADTX" to indicate tests not passing CheckTransaction()
- Fix 2 witness tests with empty vout

jl2012 committed:
Verify that verifyFlags in tx_valid and tx_invalid tests are minimal

- in tx_valid, all flags are set except those in verifyFlags
- in tx_invalid, all flags are not set except those in verifyFlags

jl2012 commented on src/test/data/tx_invalid.json:
```diff
@@ -82,174 +82,174 @@
+"010000000100010000000000000000000000000000000000000000000000000000000000000000000000ffffffff0100000000000000000000000000", "CHECKLOCKTIMEVERIFY"],
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "0"]],
-"01000000010001000000000000000000000000000000000000000000000000000000000000000000000251b1ffffffff0100000000000000000002000000", "P2SH,CHECKLOCKTIMEVERIFY"],
+"01000000010001000000000000000000000000000000000000000000000000000000000000000000000251b1ffffffff0100000000000000000002000000", "NONE"],
```
I think there is something wrong with this test, but I am not sure about its intention

jl2012 commented on src/test/data/tx_invalid.json:
```diff
@@ -82,174 +82,174 @@
+"01000000010001000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000065cd1d", "CHECKLOCKTIMEVERIFY"],
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "0"]],
-"01000000010001000000000000000000000000000000000000000000000000000000000000000000000251b100000000010000000000000000000065cd1d", "P2SH,CHECKLOCKTIMEVERIFY"],
+"01000000010001000000000000000000000000000000000000000000000000000000000000000000000251b100000000010000000000000000000065cd1d", "NONE"],
```
Another problematic test with unclear intention

practicalswift commented on src/test/transaction_tests.cpp:
```diff
@@ -238,28 +267,27 @@ BOOST_AUTO_TEST_CASE(tx_invalid)
+            for (size_t f = 1; f < verify_flags.size(); f++) {
+                unsigned int flags = verify_flags[f];
+                if (!(flags & SCRIPT_VERIFY_CLEANSTACK)) {
+                    flags &= ~SCRIPT_VERIFY_WITNESS;
```
Should the implicit conversion from negative to unsigned be made explicit here?

jl2012 replied:
Done. Also fixed a bug as I swapped SCRIPT_VERIFY_CLEANSTACK and SCRIPT_VERIFY_WITNESS here

jl2012 commented:
Added a commit to verify all validation flags are backward compatible (softfork), like #10699

jl2012 committed:
Verify that all validation flags are backward compatible

See #10699

jl2012 committed:
Apply minimal/maximal validation flags to invalid/valid script tests

- For invalid script tests:
-- reduce the number of validation flags used, to a minimally required set to fail a test
-- verify that adding flags would not validate an invalid script
- For valid script tests:
-- apply all flags except those explicitly excluded
-- verify that removing flags would not invalidate a valid script

jl2012 commented on src/test/script_tests.cpp:
```diff
@@ -155,37 +159,59 @@ CMutableTransaction BuildSpendingTransaction(const CScript& scriptSig, const CSc
-    int libconsensus_flags = flags & bitcoinconsensus_SCRIPT_FLAGS_VERIFY_ALL;
-    if (libconsensus_flags == flags) {
+    int libconsensus_flags = main_flags & bitcoinconsensus_SCRIPT_FLAGS_VERIFY_ALL;
+    if (expect || libconsensus_flags == (int)main_flags) {
```
rebased and fixed a bug here. Without `expect ||`, all valid script tests were skipped due to the use of standardness flags. If a script is valid with standardness flags, removing these flags shouldn't make it invalid.

This also fixed an existing bug that many tests containing standardness flags (mostly STRICTENC) are unnecessarily skipped.

practicalswift commented:
@jl2012 Could you rebase this PR on master and push to have Travis test the rebased version? I think there might be some UBSan runtime errors that needs to be addressed.

MarcoFalke commented:
No activity for a year. Closing for now, let me know when you want to work on this again

MarcoFalke commented:
(last commit hasn't been picked up yet)
